governance/default.go

```diff
@@ -421,6 +436,15 @@ func NewGovernance(chainConfig *params.ChainConfig, dbm database.DBManager) *Gov
 		GovernanceVotes:          NewGovernanceVotes(),
 		idxCacheLock:             new(sync.RWMutex),
 	}
+	// TODO-klaytn KIP71 default setting
+	// If DB doesn't have the KIP71 state, default values will be used
+	if g.ChainConfig.Governance == nil {
+		g.ChainConfig.Governance = &params.GovernanceConfig{}
+	}
+	if g.ChainConfig.Governance.KIP71 == nil {
+		g.ChainConfig.Governance.KIP71 = params.GetDefaultKip71Config()
+	}
+	return g
```

Comment:
While it solves the problem, It's a little hard to expect that a constructor `NewGovernance` has a side-effect. How about moving this code to `cn.New()`?

```go
chainConfig, ... := blockchain.SetupGenesisBlock(...)

// ...

// Fill defaults to missing configs
if chainConfig.Governance == nil {
    chainConfig.Governance = &params.GovernanceConfig{}
}
if chainConfig.Governance.KIP71 == nil {
      chainConfig.Governance.KIP71 = params.GetDefaultKip71Config()
}

governance := governance.NewMixedEngine(chainConfig, chainDB)

// at this point, chainConfig = (DB or genesis.json) + defaults
logger.Info("Initialised chain configuration", "config", chainConfig)
```

Reply:
I think so, but governance initialization update the params from database. so cn.New( ) always update the values to default. How do you think refactoring to split the governance initialization and update code?

Reply:
@kjeom You're right about `Governance` updating params from database e.g. in [Governance.UnmarshalJSON](https://github.com/klaytn/klaytn/blob/v1.8.3/governance/default.go#L962) via [Governance.ReadGovernanceState](https://github.com/klaytn/klaytn/blob/v1.8.3/governance/default.go#L1006). But it does not affect other chainConfig objects held by BlockChain and TxPool. That's why we should use `SetGasPrice()`, etc.

Here, new ChainConfig is created by `json.Unmarshal` and replaces existing pointer `gov.ChainConfig`. But the contents of `bc.chainConfig` won't change. They are the different pointers  after executing the `gov.ChainConfig = j.ChainConfig` line.

In contrast, adding fields to the `chainConfig` field at `NewGovernance()` do affect other code, because `g.chainConfig` and `bc.chainConfig` are the same pointer.

Reply:
I'd love to refactor `default.go` but I'm not sure how. There are so many things going on here. Instead, I would recommend to minimize modifying the code until we have a clear plan to fix it.

Reply:
Ok I got it, I will consider more about the point with no side effect.

Reply:
@blukat29 @ian0371 How do you think `governance.GetGovernance( )` to get the context in anywhere?

Reply:
Do you mean using `governance.GetGovernanceValue()` everywhere instead of pushing new values via `SetGasTarget()`?

That sounds good approach. That is similar to what I am [trying to do](https://github.com/klaytn/klaytn/pull/1381). Note that the governance part requires a lot of effort to make it pretty. Me and @ian0371 is [working on it](https://github.com/klaytn/klaytn/projects/2).

If we are going to change the way parameters are propagated, we'll have to fix for other params as well, at once. (UnitPrice, ProposerUpdateInterval, StakeUpdateInterval, ...). That should be another PR itself.

How about fitting into current structure for the sake of dynamic gas price in this PR?

Reply:
I agree that your suggestion is a better design.
However, I think it's better to leave it as is for the moment and reconsider this later.

Reply:
Thanks for your opinions. then I think it's better to leave as it is until more clear.

Reply:
@blukat29  @ian0371  I tested `cn.chainConfig's governance fields` by voting and confirmed that it works as intended (modify, persist).

1. After voting a lowerboundbasefee parameter, it exactly changed in the next epoch cycle as following capture
2. After rebooting the node, I was able to check the persistence of the parameter

- test epoch : 30
- index cache : 300 (voting confirmed)
- lowerboundbasefee : 30ston -> 250ston
![image](https://user-images.githubusercontent.com/18018028/169638639-7a5f3510-39f1-4f33-b398-1a1965bfbe71.png)

Following is my opinion that can be used as references for governance design.
- Using the `governance.chainConfig API`, we can check the chainConfig of governance. 
When the node is first booted, chainConfig instances are shared as references with `cn.chainConfig` and `governance.chainConfig`, but after rebooting, a different instance is created between the two so API results and cn.chainConfig has a difference context.  I **think we will need a clear policy on this part later.**

- For example, if `governance.chainConfig API` is used to show the value of the genesis boot, cn and governance should not share the same instance even at the genesis.
If you are looking for the context of the current node by the API, not the value of the first boot, you should always share the same instance.

Reply:
@kjeom Glad to hear that it works!

- Yes the `governance.chainConfig` is clearly flawed. There are inconsistencies among the documentation (initial chain configuration), the implementation (updated every 1024 blocks), the expectation (updated every block).
- In your use case, as you may know, be sure the use other methods including
    - console command `governance.itemsAt(klay.blockNumber)`
    - directly printing in-memory `chainConfig` variable like you did.

There are some ways to fix it.

1. Make sure `governance.chainConfig == cn.chainConfig` at all times
  a. Always share the `ChainConfig` instance throughout all codes including BlockChain, TxPool, and Governance. 
  b. Make sure Governance do not replace its `ChainConfig` pointer field.
  c. At an epoch, the changes are directly reflected to `ChainConfig`. No explicit setters (i.e. no need to `SetGasPrice`).
  d. Fix Klaytn docs to say it returns "latest chain config"

2. Or just fix `governance.chainConfig` API to return the latest value
  a. Implement `API.ChainConfig() = API.ItemsAt("latest")`
  b. Fix Klaytn docs to say it returns "latest chain config"

I'm not sure if the "initial chain config" API is really necessary. We can use `itemsAt(0)` for that.

This discussion is definitely important. I think we can continue on this topic at https://github.com/klaytn/klaytn/issues/1391.